src/quicktype-core/language/CPlusPlus.ts

```diff
@@ -971,99 +1036,209 @@ export class CPlusPlusRenderer extends ConvenienceRenderer {
+            if (recurseIntoUnion) {
+                if (nonNulls !== undefined) {
+                    for (const tt of nonNulls) {
+                        /**
+                         * We can forward declare ANYTHING in a union, however
+                         * variants we need to include directly (they are typedefs)
+                         */
+                        this.updateIncludes(maybeNull === undefined, level + 1, includes, tt, defName);
+                    }
                 }
 
+                return;
+            }
+
+            /** 
+             * Even if forward declaration is possible, 
+             * but we don't want to check every possible union
+             * member types, simple include the definition.
+             */
+            propRecord.objectType = ObjectType.Union;
+            propRecord.kind = IncludeKind.Include;
+        }
+
+        if (!propertyTypes.has(typeName)) {
+            propertyTypes.set( typeName, propRecord);
+        } else {
+            /** 
+             * We don't care what we have already as a includekind
+             * for the type, we only update this if we want to
+             * set a stronger kind
+             */
+            if (propRecord.kind === IncludeKind.Include) {
+                propertyTypes.set( typeName, propRecord);
+            }
+        }
+    }
+
+    /** 
+     * if forwardDeclare is set to true we force it, level shows how 
+     * deep we are in in the definition
+     */
+    protected updateIncludes(forwardDeclare: boolean | undefined, level:number, includes: IncludeMap, propertyType: Type, defName:string): void {
+        /** God we don't really have custom sets */
+        let propertyTypes: Map<string, IncludeRecord> = new Map();
+
+        /** 
+         * Ok if the propertyType is an union type we can SURELY
+         * forward ANY directly reachable types
+         */
+        let proposedIncludeKind:IncludeKind | undefined = undefined;
+        if (forwardDeclare !== undefined) {
+            proposedIncludeKind = forwardDeclare ? IncludeKind.ForwardDeclare : undefined;
+        } else {
+            proposedIncludeKind = propertyType instanceof UnionType || propertyType instanceof ArrayType ? IncludeKind.ForwardDeclare : undefined;
+        }
+
+        /** Recurse into ArrayType */
+        if (propertyType instanceof ArrayType) {
+            propertyType = propertyType.items;
+            const propertyTypeName = this.sourcelikeToString(this.cppType(propertyType, { needsForwardIndirection: false, needsOptionalIndirection: false, inJsonNamespace: false }, true));
+            /** We surely need to add the array to the include (if the array items are proper types) */
+            this.updatePropertyTypes(false, level+1, includes, propertyTypes, proposedIncludeKind, propertyTypeName, propertyType, defName);
+        }
+
+        directlyReachableTypes<string>(propertyType, t => {
```

Comment:
This is a bit weird - this is only run for side effects, so it doesn't have to return a set.  Ideally there would be a separate function for this, but if not, I think I think it should just return a `Set<undefined>`.

Reply:
Yeah I know, I had a though creating a new function which returns a Map or simply doesn't return anything at all, just was lazy to do it. Any particular thoughts which way to go?

Reply:
Aok, I leave you it to refactor it then, if you don't mind

Reply:
Not at all.